staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go

```diff
@@ -125,10 +126,21 @@ func withLogging(handler http.Handler, stackTracePred StacktracePred, shouldLogR
 		rl := newLoggedWithStartTime(req, w, startTime)
 		rl.StacktraceWhen(stackTracePred)
 		req = req.WithContext(context.WithValue(ctx, respLoggerContextKey, rl))
-		defer rl.Log()
+		logInRoutine := false
+		defer func() {
+			if !logInRoutine {
+				rl.Log()
+			}
+		}()
 
 		w = responsewriter.WrapForHTTP1Or2(rl)
 		handler.ServeHTTP(w, req)
+
+		// log after the task is executed in the separate routine
+		// there is a task created in handler.ServeHTTP()
```

Comment:
Let's put more context into this comment - I don't think it clarifies enough for people not familiar with it.

// We need to ensure that the request is logged after it is processed.
// In case the request is executed in a separate goroutine (created via
// WithRoutine handler in the handler chain), we want the logging to
// happen in that goroutine too, so we append it to the task.